[GEN][ZH] Add Replay Simulation feature

helmutbuhler opened the pull request:
This PR adds the ability to simulate a replay via commandline and check for crashes and mismatches.

To simulate a replay, you must place it in the Replay folder and call the game like this:
```
generalszh.exe -headless -replay "Golden Replay #1.rep"
```

The console output will be something like this:
```
Simulating Replay "Golden Replay #1.rep"
Elapsed Time: 00:13 Game Time: 10:00/99:09
Elapsed Time: 00:28 Game Time: 20:00/99:09
Elapsed Time: 00:44 Game Time: 30:00/99:09
Elapsed Time: 01:01 Game Time: 40:00/99:09
Elapsed Time: 01:29 Game Time: 50:00/99:09
Elapsed Time: 01:49 Game Time: 60:00/99:09
Elapsed Time: 02:06 Game Time: 70:00/99:09
Elapsed Time: 02:25 Game Time: 80:00/99:09
Elapsed Time: 02:59 Game Time: 90:00/99:09
Elapsed Time: 03:14 Game Time: 99:09/99:09
```

As can be seen from that output, simulating a replay with gametime 99:09 takes about 3 minutes. Replaying it with fast-forward and setting Gentool-Speed to 100x takes about 4:15. For real replays with many players the difference can be much larger.

The exit code will be 1 if an error or mismatch occures and 0 otherwise.

You can also supply `-replay` multiple times in order to simulate multiple replays one after the other.

This PR also adds another commandline `-jobs`. When this option is supplied together with multiple replays to simulate, all replays are simulated in multiple worker processes in parallel. The number after `-jobs` indicates how many processes to run in parallel. Here is an example:
```
generalszh.exe -headless -jobs 2 -replay "Golden Replay #1.rep" -replay air_air.rep -replay air_china.rep
```
This will run the first two replays in parallel, and as soon as one is finished, run the third.

Simulating replays in separate processes can speed things up a lot if you want to check many replays. But it can also help prevent mismatches because right now the game doesn't clean everything up perfectly after a game. (Eventually we will want to fix those of course)

There are a few issues with this PR:
- When calling a gui application via cmd, Windows by default hides the console output and runs the process in the background. This can be worked around with a .bat file like this:
```
START /B /W generalszh.exe -headless -replay "Golden Replay #1.rep" > sim_replay.log
echo %errorlevel%
PAUSE
```
Eventually we will want to add a cmake option to compile the game as a Console application.

~~There are two hacks at the end of `SimulateReplayListSingleProcess` to prevent crashes during gameshutdown. (It's quite possible that at least one of those will be fixed by #921, but I haven't checked yet)~~

There will be two followup PRs for this. One will add the ability to add a bunch of replays at once with a csv file. The other one will add a debug button in the ReplayMenu to simulate a replay without graphics (without having to use the commandline).

## TODO

- [x] Replicate to Generals

helmutbuhler committed:
Fix messy code and potential null pointer dereference in RecorderClass::appendNextCommand()

helmutbuhler committed:
Fix Replay playback not working after running analysis on a replay.

helmutbuhler committed:
Add replay simulation functionality

# Conflicts:
#	GeneralsMD/Code/GameEngine/Source/GameClient/GUI/GUICallbacks/Menus/ReplayMenu.cpp
#	GeneralsMD/Code/GameEngine/Source/GameLogic/System/GameLogic.cpp

helmutbuhler committed:
Add -simReplay command line option to run replays without graphics

# Conflicts:
#	GeneralsMD/Code/GameEngine/Source/Common/CommandLine.cpp

helmutbuhler committed:
Return Exitcode 1 if mismatch is detected in a replay while simulating.

# Conflicts:
#	GeneralsMD/Code/GameEngine/Source/Common/Recorder.cpp
#	GeneralsMD/Code/Main/WinMain.cpp

helmutbuhler committed:
Make Replay Simulation compile in Release and remove VERIFY_CRC.

# Conflicts:
#	GeneralsMD/Code/GameEngine/Source/GameClient/GUI/GUICallbacks/Menus/ReplayMenu.cpp

helmutbuhler committed:
Fix draw assignment in StealthUpdate::changeVisualDisguise()

helmutbuhler committed:
Show analyze and simulate button in replaymenu when DEBUG_LOGGING is defined (even in Release mode)

helmutbuhler committed:
Add code to write out replay list to textfile

# Conflicts:
#	GeneralsMD/Code/GameEngine/Source/GameClient/GUI/GUICallbacks/Menus/ReplayMenu.cpp

helmutbuhler committed:
Update WriteOutReplayList()

helmutbuhler committed:
Make Replay extra buttons show up always

helmutbuhler committed:
Add m_playbackFrameDuration to Recorder

helmutbuhler committed:
Update WriteOutReplayList and add ReadReplayListFromCsv and -simReplayList option

helmutbuhler committed:
Add more logging to SimulateReplayList and run each replay in separate process

helmutbuhler committed:
SimulateReplayList: Count errors

helmutbuhler committed:
Factor CreateProcessA into helper function

helmutbuhler committed:
Make Performance Timers work (WIP)

helmutbuhler committed:
Make SimulateReplayList faster by calling TheParticleSystemManager->reset();

helmutbuhler committed:
Add GameClient::updateHeadless()

helmutbuhler committed:
Use pipes to retrieve worker process console output

helmutbuhler committed:
Add SimulateReplayListMultiProcess. It seems to work now!

helmutbuhler committed:
Use TerminateProcess to work around crash

helmutbuhler committed:
Expand columns in WriteOutReplayList and fix ReadLineFromFile

helmutbuhler committed:
Add Job stuff to ReplayProcess

helmutbuhler committed:
Fix Jobs for VC6

helmutbuhler committed:
Revert "Make Performance Timers work (WIP)"

This reverts commit 6b23bf896710cb82e0060178472f09ac32b230c5.

helmutbuhler committed:
Merge tag 'maaaaaaaaaaaaain' into sim_replay_sh

helmutbuhler committed:
Fix date formats in comments

helmutbuhler committed:
Remove copied comment

helmutbuhler committed:
Remove bugfix that doesn't belong here

helmutbuhler committed:
Remove added dump code in RecorderClass::handleCRCMessage

helmutbuhler committed:
Remove merged headless references

helmutbuhler committed:
Fix simulation loop in ReplayMenu

helmutbuhler committed:
Make ReadReplayListFromCsv work in subfolders

helmutbuhler committed:
Make subprocesses call with -headless

helmutbuhler committed:
Allow more iniCRCs in WriteOutReplayList

helmutbuhler committed:
Move Worker Process stuff into separate file

helmutbuhler committed:
Move command creation out of WorkerProcess::StartProcess

helmutbuhler committed:
Put Replay Simulation Code into own file

helmutbuhler committed:
Put ReadReplayListFromCsv and WriteOutReplayList into separate file

helmutbuhler committed:
Add -jobs commandline

helmutbuhler committed:
Fix whitespace and revert TheFileSystem->getFileListInDirectory fix

helmutbuhler committed:
Merge branch 'main' into sim_replay_sh

# Conflicts:
#	GeneralsMD/Code/GameEngine/CMakeLists.txt
#	GeneralsMD/Code/GameEngine/Include/Common/Recorder.h
#	GeneralsMD/Code/GameEngine/Source/Common/CommandLine.cpp
#	GeneralsMD/Code/GameEngine/Source/Common/Recorder.cpp
#	GeneralsMD/Code/GameEngine/Source/GameClient/GUI/GUICallbacks/Menus/ReplayMenu.cpp
#	GeneralsMD/Code/GameEngine/Source/GameLogic/System/GameLogic.cpp
#	GeneralsMD/Code/Main/WinMain.cpp

helmutbuhler committed:
Fix wrong merge

helmutbuhler committed:
Don't check for multiple instances in headless mode

helmutbuhler committed:
Add -writeReplayList commandline and make it work with relative paths

helmutbuhler committed:
Use headless mode for -writeReplayList, -simReplay, -simReplayList. Remove -headless option

helmutbuhler committed:
Remove obsolete headless checks

helmutbuhler committed:
After using simulation button, return to scorescreen instead of Replay menu

helmutbuhler committed:
Remove obsolete headless check in GameLogic::startNewGame

helmutbuhler committed:
Remove Int mismatchFrame simplification in Recorder

helmutbuhler committed:
Remove -writeReplayList and -simReplayList commandline options (goes into other PR)

helmutbuhler committed:
Remove "Debug: Simulate Replay"-button in replaymenu (goes into other PR)

helmutbuhler commented on GeneralsMD/Code/GameEngine/Source/Common/Recorder.cpp:
```diff
@@ -1289,16 +1316,6 @@ void RecorderClass::appendNextCommand() {
 	}
 
 	GameMessage *msg = newInstance(GameMessage)(type);
```
These two changes in RecorderClass::appendNextCommand() are just simplification and have no semantic change. The original code was a bit silly and this simplifies it a lot, but I can take it out of this PR if desired. (But my hope is that it can stay here)

helmutbuhler commented on GeneralsMD/Code/GameEngine/Source/GameLogic/System/GameLogic.cpp:
```diff
@@ -3680,20 +3681,21 @@ void GameLogic::update( void )
-			msg->appendBooleanArgument( (TheRecorder && TheRecorder->getMode() == RECORDERMODETYPE_PLAYBACK) ); // playback CRC
-			DEBUG_LOG(("Appended Playback CRC on frame %d: %8.8X\n", m_frame, m_CRC));
-		}
+		bool isPlayback = (TheRecorder && TheRecorder->isPlaybackMode());
+		
+		GameMessage *msg = newInstance(GameMessage)(GameMessage::MSG_LOGIC_CRC);
+		msg->appendIntegerArgument(m_CRC);
+		msg->appendBooleanArgument(isPlayback);
+
+		// TheSuperHackers @info helmutbuhler 13/04/2025
+		// During replay simulation, we bypass TheMessageStream and instead put the CRC message
+		// directly into TheCommandList because we don't update TheMessageStream during simulation.
+		GameMessageList *messageList = TheMessageStream;
+		if (TheRecorder && TheRecorder->getMode() == RECORDERMODETYPE_SIMULATION_PLAYBACK)
+			messageList = TheCommandList;
+		messageList->appendMessage(msg);
+
+		DEBUG_LOG(("Appended %sCRC on frame %d: %8.8X\n", isPlayback ? "Playback " : "", m_frame, m_CRC));
```
Sorry if this diff is a bit hard to understand. I first deduplicated the almost identical code in the two blocks and then added some code

helmutbuhler committed:
Remove unnecessary TheControlBar != NULL check

xezon commented on GeneralsMD/Code/GameEngine/Source/Common/ReplaySimulation.cpp:
```diff
@@ -0,0 +1,186 @@
+
+			WorkerProcess p;
+			p.StartProcess(command);
+			processes.push_back(p);
```
Alternatively can do
```cpp
processes.push_back(WorkerProcess());
processes.back().StartProcess(command);
```

xezon commented on GeneralsMD/Code/GameEngine/Include/Common/Recorder.h:
```diff
@@ -48,6 +48,7 @@ class ReplayGameInfo : public GameInfo
 enum RecorderModeType CPP_11(: Int) {
 	RECORDERMODETYPE_RECORD,
 	RECORDERMODETYPE_PLAYBACK,
+	RECORDERMODETYPE_SIMULATION_PLAYBACK, // Play back replay without any graphics (TheSuperHackers @feature helmutbuhler 13/04/2025)
```
RECORDERMODETYPE_SIMULATION_PLAYBACK is used intercheably as RECORDERMODETYPE_PLAYBACK in most places. How about make the simulation just a flavour for RECORDERMODETYPE_PLAYBACK, for example with flags:

```cpp
enum RecorderModeFlags
{
	RECORDERMODETYPE_NONE = 0
	RECORDERMODETYPE_RECORD = 1 << 1,
	RECORDERMODETYPE_PLAYBACK = 1 << 2,
	RECORDERMODETYPE_HEADLESS = 1 << 3,
}
```

Then for Headless Playback combine these flags and test the flags like (m_mode & RECORDERMODETYPE_PLAYBACK) and provide convenience functions to outside:

RecorderClass::isRecording() const
RecorderClass::isPlayback() const
RecorderClass::isHeadless() const

In theory, there is also a world where a recording could be headless, for example if Machine Learning was used to play the game and record the replays.

Also, isn't `RECORDERMODETYPE_SIMULATION_PLAYBACK` synonymous for `m_mode == RECORDERMODETYPE_PLAYBACK && TheGlobalData->m_headless`? Perhaps it is not even necessary to add this flavour in this class, because headless is already queryable elsewhere.

helmutbuhler replied:
see comment below

xezon reviewed (commented):
Bunch of small things can be improved here. Overall looks solid, but I would prefer to see a separation between replay playback and headless if possible.

helmutbuhler committed:
Merge branch 'main' into sim_replay_sh

# Conflicts:
#	GeneralsMD/Code/Main/WinMain.cpp

helmutbuhler committed:
Remove refactoring out of RecorderClass::appendNextCommand()

helmutbuhler committed:
Add stupid const

helmutbuhler committed:
Rename Replay Simulation stuff

helmutbuhler committed:
Fix minor syntax stuff

helmutbuhler committed:
Use TheCommandList in RecorderClass::stopPlayback() for consistency

helmutbuhler committed:
Rename Replay Header member FrameDuration to FrameCount

helmutbuhler commented on GeneralsMD/Code/GameEngine/Include/Common/WorkerProcess.h:
```diff
@@ -0,0 +1,48 @@
+	void Update();
+
+	// Terminate Process if it's running
+	void Kill();
```
Terminate is only used in Windows, and Shutdown implicates an orderly shutdown. I know this codebase is Windows only right now, but eventually this class is probably also implemented in Linux, and kill is pretty much standard terminology for this.

helmutbuhler commented on GeneralsMD/Code/GameEngine/Source/Common/WorkerProcess.cpp:
```diff
@@ -0,0 +1,212 @@
+	// We want to make sure that when our process is killed, our workers automatically terminate as well.
+	// In Windows, the way to do this is to attach the worker to a job we own.
+	m_jobHandle = CreateJobObjectW != NULL ? CreateJobObjectW(NULL, NULL) : NULL;
+	if (m_jobHandle != NULL)
```
Yes, attaching the Job Object optional, it should still work.

helmutbuhler commented on GeneralsMD/Code/GameEngine/Source/Common/WorkerProcess.cpp:
```diff
@@ -0,0 +1,212 @@
+	{
+		// Call PeekNamedPipe to make sure ReadFile won't block
+		DWORD bytesAvailable = 0;
+		BOOL success = PeekNamedPipe(m_readHandle, NULL, 0, NULL, &bytesAvailable, NULL);
```
It's never NULL at this point (see comments above)

helmutbuhler committed:
Make methods of WorkerProcess lowercase

helmutbuhler committed:
Create helper method fetchStdOutput() in WorkerProcess

helmutbuhler commented on GeneralsMD/Code/GameEngine/Source/Common/WorkerProcess.cpp:
```diff
@@ -0,0 +1,212 @@
+		for (int i = 0; i < readBytes; i++)
+			if (buffer[i] == '\r')
+				buffer[i] = ' ';
+		buffer[readBytes] = 0;
```
I prefer 0, and it's also like this in the existing codebase

helmutbuhler committed:
Some minor changes

helmutbuhler committed:
"-simReplayList" checks belong in other PR

helmutbuhler commented on GeneralsMD/Code/Main/WinMain.cpp:
```diff
@@ -997,14 +1002,14 @@ Int APIENTRY WinMain( HINSTANCE hInstance, HINSTANCE hPrevInstance,
 		// TheSuperHackers @refactor The instance mutex now lives in its own class.
 
-		if (!rts::ClientInstance::initialize())
+		if (!headless && !rts::ClientInstance::initialize())
```
The check is necessary. Otherwise you won't be able to run the retail client while checking replays in the background. But we can clean this up in another PR. Maybe by adding a commandline option to allow multiple instances.

xezon replied:
If we run multiple clients that do not use unique client id's, then these clients will use the same user config INI files and write to the same log files.

> Otherwise you won't be able to run the retail client while checking replays in the background.

This can be solved by starting the client instance id's at 2 when headless is started. This way headless clients will never take the retail mutex.

xezon replied:
Fixed

helmutbuhler committed:
Remove out parameters in WorkerProcess::isDone

helmutbuhler committed:
variable renaming

helmutbuhler committed:
More renaming

helmutbuhler committed:
remove hacks

helmutbuhler committed:
Add SIMULATE_REPLAYS_SEQUENTIAL and some refactoring

helmutbuhler commented on GeneralsMD/Code/GameEngine/Source/Common/ReplaySimulation.cpp:
```diff
@@ -0,0 +1,186 @@
+
+	std::vector<WorkerProcess> processes;
+	int filenamePositionStarted = 0;
+	int filenamePositionDone = 0;
```
According to both C++ Core Guideslines and Googles C++ Style, it's preferable to use signed integers unless bitsemantics or modulo behavior is desired.

helmutbuhler commented on GeneralsMD/Code/GameEngine/Source/Common/ReplaySimulation.cpp:
```diff
@@ -0,0 +1,186 @@
+			filenamePositionDone++;
+		}
+
+		// Count how many processes are running
```
I don't think that would make it more readable.

helmutbuhler committed:
more minor stuff

helmutbuhler commented on GeneralsMD/Code/GameEngine/Source/Common/Recorder.cpp:
```diff
@@ -1194,17 +1215,23 @@ Bool RecorderClass::playbackFile(AsciiString filename)
 	{
-		GameMessage *msg = TheMessageStream->appendMessage( GameMessage::MSG_NEW_GAME );
+		// TheSuperHackers @info helmutbuhler 13/04/2025
+		// We send the New Game message here directly to the command list and bypass the TheMessageStream.
```
What do you mean with 'explain this on a class level'?

helmutbuhler committed:
Update comment

helmutbuhler committed:
Move new files to Core

helmutbuhler commented:
OK, I addressed most suggestions.

> Also, isn't RECORDERMODETYPE_SIMULATION_PLAYBACK synonymous for m_mode == RECORDERMODETYPE_PLAYBACK && TheGlobalData->m_headless? Perhaps it is not even necessary to add this flavour in this class, because headless is already queryable elsewhere.

Yeah, sorry, I should have mentioned why. This PR #926 will add a button to simulate a replay without graphics from the replay menu. In that case headless will be false, and we need another flag to indicate that the replay is simulated. I figured another enum member is a good and easy way to distinguish this.

Regarding the flags idea: It would work like that, but I also feel like that would be unnecessarily complicated. It won't be easy to see what flag combinations are valid and listing each possible state is just simpler. If we want to make more combinations possible in the future we can reevaluate. But I can change it if requested.

Regarding the `-headless` commandline: Combining headless mode with any other commandline option other than replay playback doesn't make sense, so I don't see the benefit of making this an explicit option. There are some commandline options (and I will add some more in followup PRs) where there is no sense to show GUI. But those options can set headless to true themself. I originally only added the option to make the first PR testable, other than that it doesn't make any sense to play the shellmap back in headless mode. The codepaths to play back a replay in headless and normal mode are also quite different, so hiding those behind the same option would be a bit misleading. I'm happy to add a `-replay` option though (in addition to the existing `-simReplay`). The original source already has some code for this that's not quite working, but I'll get it running (see `-file`).

helmutbuhler committed:
Merge branch 'main' into sim_replay_sh

# Conflicts:
#	GeneralsMD/Code/GameEngine/Source/GameClient/GUI/GUICallbacks/Menus/ReplayMenu.cpp

xezon commented on GeneralsMD/Code/GameEngine/Source/Common/WorkerProcess.cpp:
```diff
@@ -0,0 +1,212 @@
+	saAttr.bInheritHandle = TRUE;
+	HANDLE writeHandle = NULL;
+	CreatePipe(&m_readHandle, &writeHandle, &saAttr, 0);
+	SetHandleInformation(m_readHandle, HANDLE_FLAG_INHERIT, 0);
```
To me it looks like the pipe is independent of the CreateProcess call. Why would CreateProcess fail if the pipe failed? Does it fail because of the NULL `writeHandle` ?

helmutbuhler replied:
That would be my guess. But it doesn't really matter. CreatePipe doesn't fail in practice. There is no need to overengineer this.

xezon commented on GeneralsMD/Code/GameEngine/Source/Common/WorkerProcess.cpp:
```diff
@@ -0,0 +1,212 @@
+			NULL, NULL, /*bInheritHandles=*/TRUE, 0,
+			NULL, 0, &si, &pi))
+	{
+		CloseHandle(writeHandle);
```
This is a minor bug then. Cannot call CloseHandle on invalid handle. I suggest handle the error higher up, unless the Pipe is not required, then make this also work without the Pipe.

https://learn.microsoft.com/en-us/windows/win32/api/handleapi/nf-handleapi-closehandle

xezon commented on GeneralsMD/Code/GameEngine/Include/Common/WorkerProcess.h:
```diff
@@ -0,0 +1,48 @@
+
+	bool IsRunning() const;
+
+	// returns true iff the process exited.
```
uff

This does not seem to be a regular word. https://dictionary.cambridge.org/spellcheck/english/?q=iff I would not use it but ok.

Caball009 replied:
A quick search leads me to believe there are more dictionary sites that do have a reference to 'iff' as 'if and only if' than ones that do not.

xezon commented on GeneralsMD/Code/GameEngine/Source/Common/ReplaySimulation.cpp:
```diff
@@ -0,0 +1,186 @@
+
+int SimulateReplayList(const std::vector<AsciiString> &filenames, int maxProcesses)
+{
+	if (maxProcesses == -1)
```
Because it would not be necessary to have a delegate process for a single job.

xezon replied:
Fixed

xezon commented:
> OK, I addressed most suggestions.

Nice.

> Yeah, sorry, I should have mentioned why. This PR #926 will add a button to simulate a replay without graphics from the replay menu. In that case headless will be false, and we need another flag to indicate that the replay is simulated. I figured another enum member is a good and easy way to distinguish this.
> 
> Regarding the flags idea: It would work like that, but I also feel like that would be unnecessarily complicated. It won't be easy to see what flag combinations are valid and listing each possible state is just simpler. If we want to make more combinations possible in the future we can reevaluate. But I can change it if requested.

Ok makes sense.

> Regarding the `-headless` commandline: Combining headless mode with any other commandline option other than replay playback doesn't make sense, so I don't see the benefit of making this an explicit option. There are some commandline options (and I will add some more in followup PRs) where there is no sense to show GUI. But those options can set headless to true themself. I originally only added the option to make the first PR testable, other than that it doesn't make any sense to play the shellmap back in headless mode. The codepaths to play back a replay in headless and normal mode are also quite different, so hiding those behind the same option would be a bit misleading. I'm happy to add a `-replay` option though (in addition to the existing `-simReplay`). The original source already has some code for this that's not quite working, but I'll get it running (see `-file`).

There could be other use cases for headless in the future.. machine learning is one. I still think -replay and -headless are the better choice here.

| Command Line Options | What it does |
|----------------------|--------------|
| -headless               | Loads into Shellmap without a visible Client |
| -replay                 | Plays back replay(s) with a visible Client |
| -replay -headless       | Plays back replay(s) without a visible Client |
| -replay -jobs           | Plays back parallel replay(s) with multiple visible Clients |
| -replay -jobs -headless | Plays back parallel replay(s) without any visible Clients |

I think it will be good if these features are flexible and combinable like that.

helmutbuhler commented:
> What it is the plan with this change? Improve the command line arguments? Improve them later? Should I do that? Or not?

I will probably get to it this weekend. I plan to add a PR to properly fix the mutex and instance thing. I will also add the replay playback thing with graphics. I expect both to be bigger changes and didn't have time for it yet.

helmutbuhler committed:
Make -simReplay work without headless

# Conflicts:
#	GeneralsMD/Code/GameEngine/Include/Common/GlobalData.h
#	GeneralsMD/Code/Main/WinMain.cpp

helmutbuhler committed:
Add -headless option back

# Conflicts:
#	GeneralsMD/Code/Main/WinMain.cpp

helmutbuhler committed:
Rename -simReplay to -replay

helmutbuhler committed:
Fix merge

helmutbuhler commented:
> There could be other use cases for headless in the future.. machine learning is one. I still think -replay and -headless are the better choice here.
> Command Line Options 	What it does
> -headless 	Loads into Shellmap without a visible Client
> -replay 	Plays back replay(s) with a visible Client
> -replay -headless 	Plays back replay(s) without a visible Client
> -replay -jobs 	Plays back parallel replay(s) with multiple visible Clients
> -replay -jobs -headless 	Plays back parallel replay(s) without any visible Clients
> 
> I think it will be good if these features are flexible and combinable like that.

I implemented `-replay` as suggested. All mentioned combinations should work now.
After a replay is shown without `-headless` the scorescreen will appear. Pressing OK will then exit the game.
If multiple replays without `-headless` are supplied, a "Continue" button will appear in the score screen that allows the user to go to the next replay.

I still need to fix the mutex thing. For that I made #1068 which will need to be merged before this PR.

helmutbuhler committed:
Merge branch 'main' into sim_replay_sh

# Conflicts:
#	GeneralsMD/Code/GameEngine/Source/Common/CommandLine.cpp
#	GeneralsMD/Code/Main/WinMain.cpp

helmutbuhler committed:
Add m_avoidFirstInstance and m_multiInstance and make -replay work while running retail game and without RTS_MULTI_INSTANCE

helmutbuhler committed:
Merge branch 'main' of https://github.com/TheSuperHackers/GeneralsGameCode into sim_replay_sh

helmutbuhler committed:
Remove now unused GameMain( int argc, char *argv[] ) parameters

helmutbuhler committed:
Changed error handling of -replay commandline parsing to output something on the console

helmutbuhler committed:
Allow wildcards in -replay commandline

helmutbuhler committed:
Update -replay comment

helmutbuhler committed:
Use mismatchFrame to simplify printf

helmutbuhler commented:
I added some more minor things and some stuff that was requested in #925

xezon reviewed (commented):
Thank you for implementing the replay + headless. It makes this more versatile. I added a few more suggestions. I would like to see settings/states moved into `ClientInstance` and `ReplaySimulator` to make this code tidier. I tried my best to reason about it.

helmutbuhler committed:
Move m_multiInstance and m_avoidFirstInstance to ClientInstance

helmutbuhler committed:
Take out -multiInstance

helmutbuhler commented on GeneralsMD/Code/GameEngine/Include/Common/GlobalData.h:
```diff
@@ -55,6 +55,7 @@ enum AIDebugOptions CPP_11(: Int);
 // PUBLIC /////////////////////////////////////////////////////////////////////////////////////////
 
 const Int MAX_GLOBAL_LIGHTS	= 3;
+const Int SIMULATE_REPLAYS_SEQUENTIAL = -1;
```
No, it's const. It will get inlined.

xezon replied:
I can tell from studying ZH Mac built with GCC in the past that these are duplicated in every CPP.

Sample case:
```
    Line 35943: 00710498 l     O __TEXT,__literal4 _LOGICFRAMES_PER_SECONDS_REAL
    Line 36045: 00710498      d  *UND* LOGICFRAMES_PER_SECONDS_REAL:S
    Line 36097: 00710498 l     O __TEXT,__literal4 _LOGICFRAMES_PER_SECONDS_REAL
    Line 36266: 00710498      d  *UND* LOGICFRAMES_PER_SECONDS_REAL:S
    Line 36347: 00710498 l     O __TEXT,__literal4 _LOGICFRAMES_PER_SECONDS_REAL
    Line 36402: 00710498      d  *UND* LOGICFRAMES_PER_SECONDS_REAL:S
    Line 36443: 00710498 l     O __TEXT,__literal4 _LOGICFRAMES_PER_SECONDS_REAL
    Line 36573: 00710498      d  *UND* LOGICFRAMES_PER_SECONDS_REAL:S
    Line 36632: 00710498 l     O __TEXT,__literal4 _LOGICFRAMES_PER_SECONDS_REAL
    Line 36699: 00710498      d  *UND* LOGICFRAMES_PER_SECONDS_REAL:S
    Line 36755: 00710498 l     O __TEXT,__literal4 _LOGICFRAMES_PER_SECONDS_REAL
    Line 36782: 00710498      d  *UND* LOGICFRAMES_PER_SECONDS_REAL:S
    Line 36795: 00710498 l     O __TEXT,__literal4 _LOGICFRAMES_PER_SECONDS_REAL
    Line 36836: 00710498      d  *UND* LOGICFRAMES_PER_SECONDS_REAL:S
```

helmutbuhler committed:
Change comment

helmutbuhler committed:
Move CommandLine::parseCommandLineForStartup(); to DebugInit

helmutbuhler committed:
Convert one missed #if defined(RTS_MULTI_INSTANCE) to use new global

helmutbuhler commented on GeneralsMD/Code/GameEngine/Source/GameClient/ClientInstance.cpp:
```diff
@@ -31,44 +32,52 @@ bool ClientInstance::initialize()
-		}
-		s_mutexHandle = CreateMutex(NULL, FALSE, guidStr.c_str());
-		if (GetLastError() == ERROR_ALREADY_EXISTS)
+		if (TheGlobalData->m_multiInstance)
```
I agree that those globals are better placed in ClientInstance, but what you propose here is just overcomplicated.
We can just get rid of the RTS_MULTI_INSTANCE and add the -multiInstance option to simplify it further. Players with bad intentions will just be able to do it no matter what we do. And right now we don't have an easy way for testers to generate multi player replays because they don't know how to build the game themself.

I mean you are the one who made multi instance possible without broadcasts and networkcompatible. Why now not make our lifes easier using that?

helmutbuhler commented on GeneralsMD/Code/GameEngine/Source/GameLogic/System/GameLogic.cpp:
```diff
@@ -3706,20 +3706,21 @@ void GameLogic::update( void )
+		// TheSuperHackers @info helmutbuhler 13/04/2025
+		// During replay simulation, we bypass TheMessageStream and instead put the CRC message
+		// directly into TheCommandList because we don't update TheMessageStream during simulation.
+		GameMessageList *messageList = TheMessageStream;
```
I did that on purpose to save vertical space. The most basic optimizer won't care either way.

helmutbuhler committed:
spelling occured -> occurred

helmutbuhler committed:
Remove unused var

helmutbuhler commented on GeneralsMD/Code/GameEngine/Include/Common/GlobalData.h:
```diff
@@ -347,8 +351,12 @@ class GlobalData : public SubsystemInterface
+	
+	std::vector<AsciiString> m_simulateReplays; ///< If not empty, simulate this list of replays and exit. (TheSuperHackers @feature helmutbuhler 13/04/2025)
+	Int m_simulateReplayJobs; ///< Maximum number of processes to use for simulation, or SIMULATE_REPLAYS_SEQUENTIAL for sequential simulation
+	Bool m_showReplayContinueButton;
```
Let's not overengineer this. We can still move things around when we'll have to.

xezon replied:
Fixed

helmutbuhler committed:
Add comment

helmutbuhler committed:
Add console output in case map is not found

xezon commented on GeneralsMD/Code/GameEngine/Include/GameClient/ClientInstance.h:
```diff
@@ -39,6 +39,10 @@ class ClientInstance
 	static const char* getFirstInstanceName();
 
+public:
+	static Bool s_multiInstance; // Allow multiple instances of the game
```
This is wrong. These variables must not be writable after initialization is done. I have written the functions that need to be added in a previous comment. If you do not have time to do it, I can do that for you.

xezon replied:
Fixed

xezon commented on GeneralsMD/Code/GameEngine/Source/GameClient/GUI/GUICallbacks/Menus/ScoreScreen.cpp:
```diff
@@ -541,23 +541,35 @@ WindowMsgHandledType ScoreScreenSystem( GameWindow *window, UnsignedInt msg,
 				TheCampaignManager->setCampaign(AsciiString::TheEmptyString);
+				if (!TheGlobalData->m_simulateReplays.empty())
+				{
+					TheWritableGlobalData->m_showReplayContinueButton = FALSE; // Tell GameEngine to quit.
```
I do not understand this. Isn't `TheGameEngine->setQuitting` telling game engine to quit?

xezon committed:
Revert "spelling occured -> occurred" in part

This reverts commit a1266af1eb8e9908d25a979153433821a0d889ed.

xezon commented on GeneralsMD/Code/GameEngine/Include/Common/file.h:
```diff
@@ -131,15 +131,15 @@ class File : public MemoryPoolObject
 		virtual Int		seek( Int bytes, seekMode mode = CURRENT ) = NULL;	/**< Sets the file position of the next read/write operation. Returns the new file
 																																				*  position as the number of bytes from the start of the file.
-																																				*  Returns -1 if an error occured.
+																																				*  Returns -1 if an error occurred.
```
Fixed.

xezon committed:
Refactor ClientInstance code

xezon committed:
Remove GlobalData::m_showReplayContinueButton

xezon committed:
Fix SIMULATE_REPLAYS_SEQUENTIAL constexpr'ness

xezon committed:
Return WorkerProcess::startProcess if CreatePipe has failed

xezon commented on GeneralsMD/Code/GameEngine/Source/Common/ReplaySimulation.cpp:
```diff
@@ -0,0 +1,186 @@
+
+	// TheSuperHackers @todo helmutbuhler 04/05/2025
+	// Some replays cause a crash inside TheGameLogic->clearGameData().
+	// It likely has to do with Contains. We really gotta fix this, but for now we just terminate.
```
Fixed

xezon committed:
Add assert for m_readHandle

xezon committed:
Clarify -jobs range in comment

xezon committed:
Move logic to count running processes into separate function

xezon commented:
Improvements appended.

I tested Score Screen with 2 replays in the Simulator, Continue and Exit button, and it worked correctly.

ClientInstance also worked.

xezon committed:
Fix constexpr in VC6 build

xezon committed:
Fix debug compile error

xezon reviewed (approved):
From my side things look ok now.

helmutbuhler committed:
Merge branch 'main' into sim_replay_sh

helmutbuhler committed:
Auto Sync with Generals

helmutbuhler committed:
Apply manual sync for failed autosync

helmutbuhler committed:
Revert TheCommandList back to TheMessageStream in RecorderClass::stopPlayback(). This was causing the last gametime to be incorrect in console output

helmutbuhler commented:
I synced with Generals and also reverted one commit that caused the gametime in the last log to be zero.(see commit)

You can merge now.

xezon commented:
Very nice. I updated the command line arguments in the Pull Request description. I think it would be nice to have this feature description moved to our wiki.

xezon commented:
Somehow this no longer works in Windows 11:

The game launches and runs the simulation, but the output does not reach the console.

<img width="1093" height="94" alt="image" src="https://github.com/user-attachments/assets/882dbe5b-d64f-4211-b37a-9bc68f58f9f1" />

Edit: The following works in Windows 11 PowerShell Terminal:

`.\generalszh.exe -headless -replay "Golden Replay #1.rep" | Write-Output`

<img width="1136" height="198" alt="image" src="https://github.com/user-attachments/assets/66f1fd4f-333f-44f3-a22e-44de386e06fb" />

helmutbuhler commented:
@xezon: This was the behavior in Windows 10 as well:

>     * When calling a gui application via cmd, Windows by default hides the console output and runs the process in the background. This can be worked around with a .bat file like this:
> 
> 
> ```
> START /B /W generalszh.exe -headless -replay "Golden Replay #1.rep" > sim_replay.log
> echo %errorlevel%
> PAUSE
> ```
> 
> Eventually we will want to add a cmake option to compile the game as a Console application.
